File: src/settings/subtitleAppearance.ts
/** Subtitle look-and-feel mirrored from Prairie mobile/TV clients. */

export type SubtitleFontSize = "small" | "medium" | "large" | "xlarge" | "xxlarge";
export type SubtitleBackgroundStyle = "box" | "shadow" | "outline" | "none";
export type SubtitlePosition = "bottom" | "lower-third" | "top";

export interface SubtitleAppearance {
  fontSize: SubtitleFontSize;
  fontColor: string;
  backgroundColor: string;
  backgroundStyle: SubtitleBackgroundStyle;
  backgroundOpacity: number;
  textOutline: boolean;
  textOutlineColor: string;
  position: SubtitlePosition;
}

export const DEFAULT_SUBTITLE_APPEARANCE: SubtitleAppearance = {
  fontSize: "large",
  fontColor: "#ffffff",
  backgroundColor: "#000000",
  backgroundStyle: "none",
  backgroundOpacity: 75,
  textOutline: true,
  textOutlineColor: "#000000",
  position: "bottom",
};

const FONT_SIZES: SubtitleFontSize[] = ["small", "medium", "large", "xlarge", "xxlarge"];
const BG_STYLES: SubtitleBackgroundStyle[] = ["box", "shadow", "outline", "none"];
const POSITIONS: SubtitlePosition[] = ["bottom", "lower-third", "top"];

const FONT_SIZE_PX: Record<SubtitleFontSize, number> = {
  small: 28,
  medium: 36,
  large: 48,
  xlarge: 60,
  xxlarge: 72,
};

const POSITION_BOTTOM_PCT: Record<SubtitlePosition, number> = {
  top: 82,
  "lower-third": 18,
  bottom: 6,
};

function isHexColor(value: string): boolean {
  return /^#[0-9a-fA-F]{6}$/.test(value);
}

function asEnum<T extends string>(value: unknown, allowed: readonly T[], fallback: T): T {
  return typeof value === "string" && (allowed as readonly string[]).includes(value)
    ? (value as T)
    : fallback;
}

export function normalizeSubtitleAppearance(
  input: Partial<SubtitleAppearance> | null | undefined,
): SubtitleAppearance {
  const merged = { ...DEFAULT_SUBTITLE_APPEARANCE, ...input };
  const opacity =
    typeof merged.backgroundOpacity === "number" && Number.isFinite(merged.backgroundOpacity)
      ? Math.min(100, Math.max(0, Math.round(merged.backgroundOpacity)))
      : DEFAULT_SUBTITLE_APPEARANCE.backgroundOpacity;

  return {
    fontSize: asEnum(merged.fontSize, FONT_SIZES, DEFAULT_SUBTITLE_APPEARANCE.fontSize),
    fontColor: isHexColor(merged.fontColor)
      ? merged.fontColor
      : DEFAULT_SUBTITLE_APPEARANCE.fontColor,
    backgroundColor: isHexColor(merged.backgroundColor)
      ? merged.backgroundColor
      : DEFAULT_SUBTITLE_APPEARANCE.backgroundColor,
    backgroundStyle: asEnum(
      merged.backgroundStyle,
      BG_STYLES,
      DEFAULT_SUBTITLE_APPEARANCE.backgroundStyle,
    ),
    backgroundOpacity: opacity,
    textOutline: merged.textOutline === true,
    textOutlineColor: isHexColor(merged.textOutlineColor)
      ? merged.textOutlineColor
      : DEFAULT_SUBTITLE_APPEARANCE.textOutlineColor,
    position: asEnum(merged.position, POSITIONS, DEFAULT_SUBTITLE_APPEARANCE.position),
  };
}

/** CSS custom properties applied to the player host for ::cue styling. */
export function subtitleAppearanceCssVars(appearance: SubtitleAppearance): Record<string, string> {
  const normalized = normalizeSubtitleAppearance(appearance);
  const opacity = normalized.backgroundOpacity / 100;
  const bg = hexToRgba(normalized.backgroundColor, opacity);
  const fontPx = FONT_SIZE_PX[normalized.fontSize];
  const bottom = POSITION_BOTTOM_PCT[normalized.position];

  let textShadow = "none";
  // Explicit "Outline" background style wins over the textOutline toggle so the
  // picker is not shadowed by the default outline-on setting.
  if (normalized.backgroundStyle === "outline") {
    const outline = normalized.textOutlineColor;
    textShadow = [
      `1px 0 0 ${outline}`,
      `-1px 0 0 ${outline}`,
      `0 1px 0 ${outline}`,
      `0 -1px 0 ${outline}`,
    ].join(", ");
  } else if (normalized.backgroundStyle === "shadow" || normalized.textOutline) {
    const outline = normalized.textOutlineColor;
    textShadow = [
      `0 0 2px ${outline}`,
      `0 0 4px ${outline}`,
      `1px 1px 0 ${outline}`,
      `-1px -1px 0 ${outline}`,
    ].join(", ");
  }

  const background = normalized.backgroundStyle === "box" && opacity > 0 ? bg : "transparent";

  return {
    "--prairie-sub-color": normalized.fontColor,
    "--prairie-sub-bg": background,
    "--prairie-sub-size": `${fontPx}px`,
    "--prairie-sub-shadow": textShadow,
    "--prairie-sub-bottom": `${bottom}%`,
  };
}

export function hexToRgba(hex: string, alpha: number): string {
  const raw = hex.startsWith("#") ? hex.slice(1) : hex;
  const r = Number.parseInt(raw.slice(0, 2), 16);
  const g = Number.parseInt(raw.slice(2, 4), 16);
  const b = Number.parseInt(raw.slice(4, 6), 16);
  const a = Math.min(1, Math.max(0, alpha));
  return `rgba(${r}, ${g}, ${b}, ${a})`;
}

export const SUBTITLE_COLOR_CHOICES = [
  { hex: "#ffffff", label: "White" },
  { hex: "#ffff00", label: "Yellow" },
  { hex: "#00ff00", label: "Green" },
  { hex: "#00ffff", label: "Cyan" },
  { hex: "#ff00ff", label: "Magenta" },
  { hex: "#e0a84a", label: "Amber" },
] as const;

export const SUBTITLE_BG_COLOR_CHOICES = [
  { hex: "#000000", label: "Black" },
  { hex: "#141820", label: "Slate" },
  { hex: "#1a1a1a", label: "Charcoal" },
] as const;

File: flutter/packages/prairie_core/lib/src/settings/subtitle_appearance.dart
/// Mirrors src/settings/subtitleAppearance.ts, including a Flutter-side
/// equivalent of `subtitleAppearanceCssVars` for PlayerScreen's overlay
/// (`subtitleAppearanceTextStyle`/`subtitleAppearanceBottomFraction` below).
library;

import 'package:flutter/painting.dart';

enum SubtitleFontSize { small, medium, large, xlarge, xxlarge }

enum SubtitleBackgroundStyle { box, shadow, outline, none }

enum SubtitlePosition { bottom, lowerThird, top }

String subtitleFontSizeWire(SubtitleFontSize v) => v.name;
String subtitleBackgroundStyleWire(SubtitleBackgroundStyle v) => v.name;
String subtitlePositionWire(SubtitlePosition v) => switch (v) {
  SubtitlePosition.lowerThird => 'lower-third',
  _ => v.name,
};

T _asEnum<T>(String? value, List<T> allowed, T fallback, String Function(T) wire) {
  if (value == null) return fallback;
  for (final option in allowed) {
    if (wire(option) == value) return option;
  }
  return fallback;
}

bool _isHexColor(String? value) => value != null && RegExp(r'^#[0-9a-fA-F]{6}$').hasMatch(value);

class SubtitleAppearance {
  const SubtitleAppearance({
    this.fontSize = SubtitleFontSize.large,
    this.fontColor = '#ffffff',
    this.backgroundColor = '#000000',
    this.backgroundStyle = SubtitleBackgroundStyle.none,
    this.backgroundOpacity = 75,
    this.textOutline = true,
    this.textOutlineColor = '#000000',
    this.position = SubtitlePosition.bottom,
  });

  final SubtitleFontSize fontSize;
  final String fontColor;
  final String backgroundColor;
  final SubtitleBackgroundStyle backgroundStyle;
  final int backgroundOpacity;
  final bool textOutline;
  final String textOutlineColor;
  final SubtitlePosition position;

  SubtitleAppearance copyWith({
    SubtitleFontSize? fontSize,
    String? fontColor,
    String? backgroundColor,
    SubtitleBackgroundStyle? backgroundStyle,
    int? backgroundOpacity,
    bool? textOutline,
    String? textOutlineColor,
    SubtitlePosition? position,
  }) => SubtitleAppearance(
    fontSize: fontSize ?? this.fontSize,
    fontColor: fontColor ?? this.fontColor,
    backgroundColor: backgroundColor ?? this.backgroundColor,
    backgroundStyle: backgroundStyle ?? this.backgroundStyle,
    backgroundOpacity: backgroundOpacity ?? this.backgroundOpacity,
    textOutline: textOutline ?? this.textOutline,
    textOutlineColor: textOutlineColor ?? this.textOutlineColor,
    position: position ?? this.position,
  );

  Map<String, dynamic> toJson() => {
    'fontSize': subtitleFontSizeWire(fontSize),
    'fontColor': fontColor,
    'backgroundColor': backgroundColor,
    'backgroundStyle': subtitleBackgroundStyleWire(backgroundStyle),
    'backgroundOpacity': backgroundOpacity,
    'textOutline': textOutline,
    'textOutlineColor': textOutlineColor,
    'position': subtitlePositionWire(position),
  };

  factory SubtitleAppearance.fromJson(Map<String, dynamic>? json) {
    const fallback = SubtitleAppearance();
    if (json == null) return fallback;
    final opacity = json['backgroundOpacity'];
    return SubtitleAppearance(
      fontSize: _asEnum(json['fontSize'] as String?, SubtitleFontSize.values, fallback.fontSize, subtitleFontSizeWire),
      fontColor: _isHexColor(json['fontColor'] as String?) ? json['fontColor'] as String : fallback.fontColor,
      backgroundColor: _isHexColor(json['backgroundColor'] as String?) ? json['backgroundColor'] as String : fallback.backgroundColor,
      backgroundStyle: _asEnum(json['backgroundStyle'] as String?, SubtitleBackgroundStyle.values, fallback.backgroundStyle, subtitleBackgroundStyleWire),
      backgroundOpacity: opacity is num ? opacity.round().clamp(0, 100) : fallback.backgroundOpacity,
      textOutline: json['textOutline'] == true,
      textOutlineColor: _isHexColor(json['textOutlineColor'] as String?) ? json['textOutlineColor'] as String : fallback.textOutlineColor,
      position: _asEnum(json['position'] as String?, SubtitlePosition.values, fallback.position, subtitlePositionWire),
    );
  }
}

/// Mirrors `SUBTITLE_COLOR_CHOICES`/`SUBTITLE_BG_COLOR_CHOICES`.
const subtitleColorChoices = [
  ('#ffffff', 'White'),
  ('#ffff00', 'Yellow'),
  ('#00ff00', 'Green'),
  ('#00ffff', 'Cyan'),
  ('#ff00ff', 'Magenta'),
  ('#e0a84a', 'Amber'),
];

const subtitleBgColorChoices = [
  ('#000000', 'Black'),
  ('#141820', 'Slate'),
  ('#1a1a1a', 'Charcoal'),
];

const _fontSizePx = {
  SubtitleFontSize.small: 20.0,
  SubtitleFontSize.medium: 24.0,
  SubtitleFontSize.large: 30.0,
  SubtitleFontSize.xlarge: 38.0,
  SubtitleFontSize.xxlarge: 46.0,
};

const _positionBottomFraction = {
  SubtitlePosition.top: 0.82,
  SubtitlePosition.lowerThird: 0.18,
  SubtitlePosition.bottom: 0.06,
};

Color _colorFromHex(String hex) => Color(int.parse('FF${hex.substring(1)}', radix: 16));

/// Fraction of the player surface's height, measured from the bottom, that
/// the caption should sit at. Mirrors `POSITION_BOTTOM_PCT`.
double subtitleAppearanceBottomFraction(SubtitleAppearance appearance) => _positionBottomFraction[appearance.position]!;

/// Text style for the caption. Mirrors the text half of
/// `subtitleAppearanceCssVars` (outline via multi-directional shadows).
TextStyle subtitleAppearanceTextStyle(SubtitleAppearance appearance) {
  final outline = _colorFromHex(appearance.textOutlineColor);
  final wantsOutline = appearance.backgroundStyle == SubtitleBackgroundStyle.outline || appearance.textOutline;
  return TextStyle(
    color: _colorFromHex(appearance.fontColor),
    fontSize: _fontSizePx[appearance.fontSize],
    fontWeight: FontWeight.w600,
    shadows: wantsOutline
        ? [
            Shadow(color: outline, offset: const Offset(1, 0)),
            Shadow(color: outline, offset: const Offset(-1, 0)),
            Shadow(color: outline, offset: const Offset(0, 1)),
            Shadow(color: outline, offset: const Offset(0, -1)),
          ]
        : null,
  );
}

/// Background decoration behind the caption text, or `null` for none.
/// Mirrors the `backgroundStyle === "box"` branch of `subtitleAppearanceCssVars`.
Color? subtitleAppearanceBackgroundColor(SubtitleAppearance appearance) {
  if (appearance.backgroundStyle != SubtitleBackgroundStyle.box) return null;
  final opacity = appearance.backgroundOpacity / 100;
  if (opacity <= 0) return null;
  return _colorFromHex(appearance.backgroundColor).withValues(alpha: opacity);
}
